### PR TITLE
Rooms collector fixes

### PR DESCRIPTION
**Pull request by Kubsy**
branch by Monty

- Optimizes room collecting performance 
- Fixes disappearing rooms in certain cases 

PLEASE ALSO TEST IF YOU CAN MOVE AND SCALE STATICS VIA LUA ON THIS PR!!!

### CODE REVIEW

File: TombEngine/Game/control/control.cpp [hunk 1]
```diff
@@ -94,6 +96,8 @@ short NextItemFree;
 short NextFxActive;
 short NextFxFree;
 
+int ControlPhaseTime;
```
Comment (Lwmte) [inline TombEngine/Game/control/control.cpp:99]: Please remove debug values or move it to another file. Preferably, we need to have separate `debug.cpp/.h` files from which you will call frame time counters.

File: TombEngine/Game/control/control.cpp [hunk 2]
```diff
@@ -103,6 +107,8 @@ int DrawPhase()
 
 GameStatus ControlPhase(int numFrames, bool demoMode)
 {
+	auto time1 = std::chrono::high_resolution_clock::now();
```
Comment (Lwmte) [inline TombEngine/Game/control/control.cpp:110]: As above.

File: TombEngine/Game/control/control.cpp [hunk 3]
```diff
@@ -317,6 +323,12 @@ GameStatus ControlPhase(int numFrames, bool demoMode)
 		}
 	}
 
+	using ns = std::chrono::nanoseconds;
```
Comment (Lwmte) [inline TombEngine/Game/control/control.cpp:326]: As above. Please don't spam `ControlPhase` with debug processing.

File: TombEngine/Game/room.h [hunk 4]
```diff
@@ -47,6 +47,7 @@ struct MESH_INFO
 	Vector4 color;
 	short HitPoints;
 	std::string luaName;
+	bool Dirty;
```
Comment (Lwmte) [inline TombEngine/Game/room.h:50]: What is the meaning of this flag?

File: TombEngine/Renderer/Renderer11.cpp [hunk 5]
```diff
@@ -495,9 +495,9 @@ namespace TEN::Renderer
 	{
 		D3D11_RECT rects;
 		rects.left = s.left;
-		rects.top = m_screenHeight - s.top;
```
Comment (Lwmte) [inline TombEngine/Renderer/Renderer11.cpp:498]: Were scissor tests successful after that change? I remember doing the same "fix" you did there, but it badly messed tests and made them work even more incorrectly.

File: TombEngine/Renderer/Renderer11.h [hunk 6]
```diff
@@ -90,11 +91,26 @@ namespace TEN::Renderer
...
 		int ObjectNumber;
 		int RoomNumber;
-		Vector3 Position;
+		int IndexInRoom;
+		Pose Pose;
 		Matrix World;
 		Vector4 Color;
 		Vector4 AmbientLight;
 		std::vector<RendererLight*> LightsToDraw;
+		std::vector<RendererLight*> CachedRoomsLights;
```
Comment (Lwmte) [inline TombEngine/Renderer/Renderer11.h:100]: Please rename this vector to `CachedRoomLights`. In English, you don't need to make previous word plural, if last one is also plural.

File: TombEngine/Renderer/Renderer11Compatibility.cpp [hunk 7]
```diff
@@ -168,13 +168,63 @@ namespace TEN::Renderer
 			r->EffectsToDraw.reserve(MAX_ITEMS_DRAW);
 			r->TransparentFacesToDraw.reserve(MAX_TRANSPARENT_FACES_PER_ROOM);
 			
+			Vector3 boxMin = Vector3(room.x + WALL_SIZE, room.maxceiling - STEP_SIZE, room.z + WALL_SIZE);
+			Vector3 boxMax = Vector3(room.x + (room.xSize - 1) * WALL_SIZE, room.minfloor + STEP_SIZE, room.z + (room.zSize - 1) * WALL_SIZE);
+			Vector3 centre = (boxMin + boxMax) / 2.0f;
+			Vector3 extens = boxMax - centre;
```
Comment (Lwmte) [inline TombEngine/Renderer/Renderer11Compatibility.cpp:174]: `extents`, not `extens`.

File: TombEngine/Renderer/Renderer11Compatibility.cpp [hunk 8]
```diff
@@ -168,13 +168,63 @@ namespace TEN::Renderer
...
+			r->BoundingBox = BoundingBox(centre, extens);
```
Comment (Lwmte) [inline TombEngine/Renderer/Renderer11Compatibility.cpp:175]: `center`, not `centre`.

File: TombEngine/Renderer/Renderer11Draw.cpp [hunk 9]
```diff
@@ -2002,7 +2008,13 @@ namespace TEN::Renderer
 			BindConstantBufferVS(CB_ROOM, m_cbRoom.get());
 			BindConstantBufferPS(CB_ROOM, m_cbRoom.get());
 
-			SetScissor(room->Clip);
+			RendererRectangle rect;
```
Comment (Lwmte) [inline TombEngine/Renderer/Renderer11Draw.cpp:2011]: Can we maybe move this calculation to `RendererRoom` subroutine or something?

File: TombEngine/Renderer/Renderer11Frame.cpp [hunk 10]
```diff
@@ -473,112 +399,153 @@ namespace TEN::Renderer
...
 			light.LocalIntensity = intensity;
 			light.Distance = distance;
 
 			tempLights.push_back(&light);
 		}
-
-		// Check current room and also neighbour rooms
-		for (int roomToCheck : room.Neighbors)
+	
+		if (!useCachedRoomsLights)
```
Comment (Lwmte) [inline TombEngine/Renderer/Renderer11Frame.cpp:430]: `useCachedRoomLights`, not `useCachedRoomsLights`.

**Comment by Lwmte** [comment]
I recommend to thoroughly test how lighting works now, and if light model is consistent. It seems light collector was also modified along with room collector, so this must be tested.

**Comment by Kubsy** [comment]
Tested, everything works correctly.

File: TombEngine/Game/control/control.cpp (hunk 1 above)
Comment (MontyTRC89) [reply TombEngine/Game/control/control.cpp:99]: Yeah it's not optimal. But in the next release I would implement more timings from control phase and so your suggestion is valid.

File: TombEngine/Renderer/Renderer11.cpp (hunk 5 above)
Comment (MontyTRC89) [reply TombEngine/Renderer/Renderer11.cpp:498]: From test it seems fine. Now rooms collector works in clip space, so -1 is bottom of screen that's why we need to do this operation. I will polish this a bit before the release.

File: TombEngine/Renderer/Renderer11.h (hunk 6 above)
Comment (MontyTRC89) [reply TombEngine/Renderer/Renderer11.h:100]: I am a donkey in english sometimes :-D agreed

File: TombEngine/Renderer/Renderer11Compatibility.cpp (hunk 8 above)
Comment (MontyTRC89) [reply TombEngine/Renderer/Renderer11Compatibility.cpp:175]: Ok!

File: TombEngine/Renderer/Renderer11Draw.cpp (hunk 9 above)
Comment (MontyTRC89) [reply TombEngine/Renderer/Renderer11Draw.cpp:2011]: It's part of my polish intentions like I have said about the scissor rect.

**Comment by Kubsy** [comment]
https://youtu.be/84NI4d0Ycbk from geckokid,

File: TombEngine/TombEngine.vcxproj [hunk 11]
```diff
@@ -23,7 +23,7 @@
     <ConfigurationType>Application</ConfigurationType>
     <UseDebugLibraries>true</UseDebugLibraries>
     <CharacterSet>MultiByte</CharacterSet>
-    <PlatformToolset>v142</PlatformToolset>
+    <PlatformToolset>v143</PlatformToolset>
```
Comment (TokyoSU) [inline TombEngine/TombEngine.vcxproj:26]: i don't think @Lwmte will let this pass :x

File: TombEngine/TombEngine.vcxproj [hunk 12]
```diff
@@ -1163,13 +1163,13 @@ CALL gen.bat</Command>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets">
     <Import Project="..\packages\Microsoft.XAudio2.Redist.1.2.8\build\native\Microsoft.XAudio2.Redist.targets" Condition="Exists('..\packages\Microsoft.XAudio2.Redist.1.2.8\build\native\Microsoft.XAudio2.Redist.targets')" />
-    <Import Project="..\packages\directxtk_desktop_2017.2021.11.8.1\build\native\directxtk_desktop_2017.targets" Condition="Exists('..\packages\directxtk_desktop_2017.2021.11.8.1\build\native\directxtk_desktop_2017.targets')" />
+    <Import Project="..\packages\directxtk_desktop_2017.2022.3.24.2\build\native\directxtk_desktop_2017.targets" Condition="Exists('..\packages\directxtk_desktop_2017.2022.3.24.2\build\native\directxtk_desktop_2017.targets')" />
```
Comment (TokyoSU) [inline TombEngine/TombEngine.vcxproj:1169]: Same as above, well i would like to update it to 2019 instead... since 2017 is deprecated.

**Comment by Adngel** [comment]
I could play my 3 levels with not crashes.

In the enemy level there was a room caused the others dissapeared a lot, with branch I'm not experiencing that. (It was a small diagonal room in a diagonal corridor, a bit unusual anyway).

**Pull request merged**